lib/ood_core/batch_connect/template.rb

```diff
@@ -121,7 +121,17 @@ def bash_helpers
                 port_used () {
                   local port="${1#*:}"
                   local host=$((expr "${1}" : '\\(.*\\):' || echo "localhost") | awk 'END{print $NF}')
-                  nc -w 2 "${host}" "${port}" < /dev/null &> /dev/null
+                  if command -v nc >/dev/null 2>&1; then
+                    nc -w 2 "${host}" "${port}" < /dev/null > /dev/null 2>&1
+                  elif command -v lsof >/dev/null 2>&1; then
+                    lsof -i :"${port}" >/dev/null 2>&1
+                  elif [ -n "$SHELL" ] && [ "$SHELL" = "/bin/bash" ]; then
```

Comment:
Eric expressed concern about `/dev/tcp` not being available for every version of Bash. Is that concern valid? If so this does not test to ensure that the version of Bash is compatible with `/dev/tcp`.

Reply:
Not sure, I just checked on the 3.0 container and it existed there and 3.0 is from 2004.  We could make stronger checks.  It would throw `No such file or directory` if it didn't exist.

Reply:
/dev/tcp needs to be checked for sure. AFAIK centos6 bash doesn’t have this even though we dropped support for centos6 on the webnode.

Also these check for the existence of the command and the assume any non zero status of the command means the port is not in use. This is also a problem. We should check for the specific exit status that indicates the port in use or not. If another exit status occurs, we should assume that means the method being used will not work and move onto the next method to check.

An example of the importance of this is VT where nc exists as a command but when users execute it the stderr gets the output something like “You do not have permission to run this command”.

Reply:
For testing Bash how about this:

```bash
function can_use_bash() {
    ( exec 3<>/dev/tcp/www.google.com/80 ) 2>/dev/null
    "$?"
}
```

Reply:
Yea i keep forgetting this is open.  

I'm not sure if we can get the level of error handling through return codes.  I just tried a bad device `/dev/tcp1` and it returned 1, then I tried an unknown/bad host `/dev/tcp/www.jeffohrstrom.biz/80` and got 1 again. 

I did get different error messages, and that's probably what we'll have to go on, actually grepping for `No such file or directory` and `Connection refused|Invalid argument` (`/dev/tcp/localhost/80` gave me connection refused).

Reply:
That might make more sense. I don't love pointing at Google just to test if Bash is configured to use TCP. I'd love some way to just ask Bash if it supports a given feature. There's (`readelf`)[https://linux.die.net/man/1/readelf], which we could use like this: `readelf -a /usr/bin/bash | grep have_devfd`, but I am not sure that `have_devfd` is the correct symbol to look for. It seems promising...

Reply:
I think we may have a winner: https://github.com/bminor/bash/blob/64447609994bfddeef1061948022c074093e9a9f/lib/sh/netopen.c#L1-L3

So the `readelf` call should really be: `readelf -a /usr/bin/bash | grep netopen`

Reply:
Judging by the binary produced by the source available from GNU (http://ftp.gnu.org/gnu/bash/) it appears that `--enable-net-redirections` is the default for Bash. In Bash 5.x the configuration option ` --disable-net-redirections` is needed to produce Bash without `/dev/tcp/*/*` capability.

I have confirmed that the feature can be detected in a Bash binary using: `strings /bin/bash | grep tcp`. `strings` is present in base CentOS containers when even `which` is not, and so seems like a safe dependency.